Issue 290 - add Lists/Arrays/Hashes for Python3

TomReilly1 opened the pull request:
<!-- Hello! Thank you for contributing! -->

<!-- Please do NOT modify the headers (the ## lines). Just add or change the content in between them. -->


## What GitHub issue does this PR apply to?

Resolves #290 


## What changed and why?

Created the 'lists.json' file to provide concepts for the Lists/Arrays/Hashes structure for Python 3.


## Checklist

<!-- Either add an X inside the [ ], or submit the PR and click the checkboxes. -->

- [X] I claimed any associated issue(s) and they are not someone else's
- [X] I have looked at documentation to ensure I made any revisions correctly
- [X] I tested my changes locally to ensure they work
- [ ] (If editing Django) I have added or edited any appropriate unit tests for my changes


## Any additional comments or things to be aware of while reviewing?

<!-- Please replace this line with any comments -->

TomReilly1 committed:
first commit for python3 lists

TomReilly1 committed:
removed ':names' fields

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+        "list_name.insert(0, value)"
+      ]
+    },
+    "insert_into_end_of_ordered_mutable_list": {
+      "code": [
+        "list_name.append(value)"
+      ]
+    },
+    "insert_into_middle_of_ordered_mutable_list": {
+      "code": [
+        "list_name.insert(index, value)"
+      ]
+    },
+    "erase_element_at_beginning_of_ordered_mutable_list": {
+      "code": [
+        "list_name.pop(0)"
+      ]
+    },
+    "erase_element_at_end_of_ordered_mutable_list": {
+      "code": [
+        "list_name.pop()"
+      ]
+    },
+    "erase_element_in_middle_of_ordered_mutable_list": {
+      "code": [
+        "list_name.pop(index)",
+        "list_name.remove(value)"
+      ],
+      "comment": "First line removes element by index. Second line removes element by value"
+    },
+    "swap_elements_in_ordered_mutable_list": {
+      "code": [
+        "list_name[index1], list_name[index2] = list_name[index2], list_name[index1]"
+      ]
+    },
+    "delete_ordered_mutable_list": {
+      "code": [
+        "del list_name"
+      ]
+    },
```
These (L 9-83) should be in `*unordered_list*`, i don't think python provides ordered lists.
Not sure if it's me or sth but github seems to delete my line number annotations so I'm putting them explicitly.

TomReilly1 replied:
Python lists are ordered in a sense that its elements have a specific order, unless CT's definition of ordered is different than that.

cafce25 replied:
You're right. CT uses it inconsistently at the moment and I've created #481. Let's decide one way or the other over in the issue, I think leave it here then and get merged in the meantime.

geekygirlsarah replied:
The definition of "ordered" that I was thinking of when I added lists corresponds to Python's lists being ordered lists.

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+      "code": [
+        "dict_name = {}",
+        "dict_name = {key: value}",
+        "dict_name = {key1: value1, key2: value2, ...}"
+      ],
+      "comment": "First line creates an empty dict. Second line creates a dict with one key/value pair. Third line creates a dict with multiple key/value pairs"
+    },
+    "get_key_from_mutable_set": {
+      "code": [
+        "list(dict_name.keys())[index]"
+      ]
+    },
+    "get_value_from_mutable_set": {
+      "code": [
+        "dict_name[key]",
+        "dict_name.get(key)",
+        "dict_name.get(key, default_value)"
+      ],
+      "comment": "If the key cannot be found, .get(key) returns a default value (or None if not specified), whereas [key] will raise an exception"
+    },
+    "get_all_keys_from_mutable_set": {
+      "code": [
+        "dict_name.keys()"
+      ]
+    },
+    "get_all_values_from_mutable_set": {
+      "code": [
+        "dict_name.values()"
+      ]
+    },
+    "swap_key_and_value_in_mutable_set": {
+      "code": [
+        "dict_name = dict([(value, key) for key, value in dict_name.items()])"
+      ]
+    },
+    "delete_mutable_set": {
+      "code": [
+        "del dict_name"
+      ]
+    },
```
Python `dict`s are what codethesaurus calls `hashed lists` and should be in the `*mutable_hashed_list*` concepts.
L 282-322

cafce25 replied:
On rereading this seems I mixed it up here, of course `*set*` is the right place for these.

geekygirlsarah replied:
This sounds resolved?

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+      "code": [
+        "del set_name"
+      ]
+    },
```
Above (L 9-83) concepts should go here, these (L 84 - 165) are set operations and belong in the `*set*` definitions below.

cafce25 replied:
Same as above, leave it here, decide later.

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+        "del set_name"
+      ]
+    },
+    "name_of_ordered_immutable_list": {
+      "code": [
+        "tuple"
+      ]
+    },
+    "create_a_ordered_immutable_list": {
+      "code": [
+        "tuple_name = ()",
+        "tuple_name(value,)",
+        "tuple_name(value1, value2, ...)"
+      ],
+      "comment": "First line creates an empty tuple. Second line creates a tuple with one value. Third line creates a tuple with multiple values"
+    },
+    "ordered_immutable_list_start_number": {
+      "code": [
+        "0"
+      ]
+    },
+    "access_element_in_ordered_immutable_list": {
+      "code": [
+        "tuple_name[index]"
+      ]
+    },
+    "delete_ordered_immutable_list": {
+      "code": [
+        "del tuple_name"
+      ]
+    },
```
I'm not sure about putting tuples as immutable lists, sure in python you can treat them as such but in my mind a tuple is fundamentally different from a list.
To me lists are collection of an unspecified amount of same/similar elements,
while tuples are a often heterogenous collection of a fixed amount of items.

TomReilly1 replied:
I think the biggest difference between a Python list and a Python tuple is mutability, which naturally leads to the former being dynamic and the latter being fixed. In terms of of heterogeneous vs. homogeneous, both Python lists and Python tuples are capable of storing either type of collections, even if one makes more sense for the other.

I am not too familiar with the differences between tuples and immutable lists from other languages, but if the main difference is what data each contains (heterogeneous vs. homogeneous), I think Python tuples could be classified as both a tuple and immutable list.

cafce25 replied:
Hmm yes, I thought of it as either or, but it fit's both so why not add it in both places (when we add tuples)!

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+        "value in list_name"
+      ],
+      "comment": "returns a True/False value"
+    },
```
I believe this should be `list_name.index(value)`

TomReilly1 replied:
Could have both since they both find by value. Would comment that one returns a boolean and the other returns an index/integer

cafce25 commented on web/thesauruses/python/3/lists.json:
```diff
@@ -0,0 +1,492 @@
+    "unordered_mutable_list_start_number": {
+      "code": [
+        ""
+      ],
```
These empty code blocks for unimplemented concepts can be removed.

TomReilly1 replied:
Will do

cafce25 reviewed (changes requested):
Hey, thanks for taking the time to write this up, very much appreciated!

you mixed up some of the concept <-> sample mappings, (which isn't hard cause all sorts of languages call the different concepts all sort of things)
in summary I think:
- everything under \*ordered_list* should be in \*unordered_list*
- everything under \*unordered_list* should be in \*set*
- everything under \*set* should be in \*hashed_list*
- all the things in \*ordered_list* should be declared unimplemented.
- tuple might be a candidate for new concepts instead of trying to fit it into immutable ordered lists (i can think of a couple of languages that have them & immutable lists).

Other than that I think you did a great job.

geekygirlsarah committed:
Merge branch 'main' into issue-290-add-python3-lists

geekygirlsarah reviewed (changes requested):
I think this is pretty good. Sounds like cafce25 found a couple of things to fix, but I can probably merge in after that!

geekygirlsarah committed:
Merge branch 'main' into issue-290-add-python3-lists

TomReilly1 committed:
removed empty code

TomReilly1 committed:
added another method for 'find_element_by_value'

TomReilly1 committed:
fix grammer for comment

TomReilly1 committed:
Merge branch 'codethesaurus:main' into issue-290-add-python3-lists

TomReilly1 committed:
Merge branch 'issue-290-add-python3-lists' of github.com:TomReilly1/codethesaur.us into issue-290-add-python3-lists

TomReilly1 commented:
OK I think we are good now. Removed the empty code blocks and added the method for finding an element by value.
I left the lists, sets, and tuples as they were.

geekygirlsarah committed:
Merge branch 'main' into issue-290-add-python3-lists

geekygirlsarah commented:
Thank you for working on this and for all the changes! I'll merge it in.
